lm_scorer/models/gpt2.py

```diff
@@ -27,47 +43,41 @@ def _add_special_tokens(self, text: str) -> str:
     def _tokens_log_prob_for_batch(
         self, text: List[str]
     ) -> List[Tuple[torch.DoubleTensor, torch.LongTensor, List[str]]]:
-        device = self.model.device
-
-        outputs = []
-        for sentence in text:
-            # len(sentence) = seq_len + 2
-            sentence = self._add_special_tokens(sentence)
-            # len(tokens) = seq_len + 2
-            tokens: List[str] = self.tokenizer.tokenize(sentence)
-            # ids.shape = [1, seq_len + 2]
-            ids = torch.tensor(  # pylint: disable=not-callable
-                [self.tokenizer.convert_tokens_to_ids(tokens)],
-                device=device,
-                dtype=torch.long,
-            )
-
-            with torch.no_grad():
-                model_outputs: Tuple[torch.Tensor] = self.model(ids)
-
-            # pred_scores.shape = [1, seq_len + 2, vocab_size]
-            pred_scores = model_outputs[0].double()
-
-            # len(tokens) = seq_len + 1
-            tokens = tokens[1:]
-            # ids.shape = [1, seq_len + 1, vocab_size]
-            ids = ids[:, 1:]
-            # pred_scores.shape = [1, seq_len + 1, vocab_size]
-            pred_scores = pred_scores[:, :-1, :]
-
-            # ids_scores.shape = [1, seq_len + 1]
-            ids_scores = pred_scores.gather(2, ids.unsqueeze(2)).squeeze(2)
-            # log_prob.shape = [1, seq_len + 1]
-            log_probs = ids_scores - pred_scores.logsumexp(2)
-
-            log_probs = log_probs.squeeze(0)
-            ids = ids.squeeze(0)
-
-            log_probs = cast(torch.DoubleTensor, log_probs)
-            ids = cast(torch.LongTensor, ids)
-
-            output = (log_probs, ids, tokens)
+        outputs: List[Tuple[torch.DoubleTensor, torch.LongTensor, List[str]]] = []
+        if len(text) == 0:
+            return outputs
+
+        # TODO: Handle overflowing elements for long sentences
+        text = list(map(self._add_special_tokens, text))
+        encoding: BatchEncoding = self.tokenizer.batch_encode_plus(
+            text, return_tensors="pt",
+        )
+        with torch.no_grad():
+            ids = encoding["input_ids"].to(self.model.device)
+            nopad_mask = ids != self.tokenizer.pad_token_id
+            logits: torch.Tensor = self.model(ids)[0]
+
+        for sent_index in range(len(text)):
+            sent_nopad_mask = nopad_mask[sent_index]
+            # len(tokens) = len(text[sent_index]) + 1
+            first_pad_index = int(torch.sum(sent_nopad_mask).item())
+            sent_tokens = encoding.tokens(sent_index)[1:first_pad_index]
```

Comment:
```suggestion
            sent_tokens = [t for i,t in enumerate(encoding.tokens(sent_index)) if sent_nopad_mask[i]]
```

Reply:
Hm, I did it because I try to avoid for loop if possible. In fact in the version I am currently using for my project, I also removed the main for loop and handle the full output directly. I did not proposed that here because it was less convenient here because with the return of the sent tokens.  
But ok, we could also do that if you prefer.

Reply:
Does that for loop impact the performance so much?
The point of the for loop is that there might be models that add pading on the left instead.
It is not the case for GPT2, but I would like the code to be general enough to be easily used for other models as well.

Reply:
Yes, I also had the same thought about left padding did it because we were in gpt2.py and that we control the batch_encode_plus call, but I am ok with your point and I will change that. 
`Does that for loop impact the performance so much?
`
I don't know, for now I think not so much because for now much of time is spent on the forward pass (maybe in the futur where model inference will be much quicker).